src/blog/speeding-up-v8-heap-snapshots.md

```diff
@@ -0,0 +1,161 @@
+
+Things were… not right: the offset of many items was over 20 entries. But some cases were really bad, in the order of thousands of positions off from their expected position!
+
+Part of the problem was caused by the scripts using strings for lots of numbers - and especially a big range of numbers from 1 to several hundreds without gaps. The hash key algorithm had two implementations, one for numbers and another for other strings. While the string hash function was quite classical, the problem came with the numbers implementation that would basically be the value of the number - with a prefix for the number of digits.
+
+Some examples of problems with this hash function:
+
+- Once we had strings with a hash key value in lower positions, then the storing of new numbers would offset all the positions.
+- Or even worse: if we would introduce a string with a hash key value that would be a low number, and we already found several hundreds or thousands of consecutive numbers, then that value would be moved several hundreds or thousands of positions.
+
+What did I do to fix it? As the problem comes mostly from numbers represented as strings that would fall in consecutive positions, I modified the hash function so we would rotate the resulting hash value 2 positions to the left. So, for each number, we would introduce 3 free positions. Why 2? Empirical testing across several work-sets showed this number was the best choice to minimize collisions.
+
+[This hashing fix](https://chromium-review.googlesource.com/c/v8/v8/+/4428811) has landed in V8.
+
+## Second Optimization: Caching source positions
+
+After fixing the hashing, we re-profiled and found a further optimization opportunity that would take a significant part of the time. *An old friend*.
+
+For each allocation in the V8 heap, the heap snapshot tries to record the call stack responsible for the allocation. Therefore, for each stack frame, it needs to know the function name and its source location (filename, line number, column number). It was the fetching of this information that turned out to be super slow!
+
+What was happening? It was something similar to [what I fixed in the ETW stack walk](https://chromium-review.googlesource.com/c/v8/v8/+/3669698) and that I explained in [this post](https://blogs.igalia.com/dape/2022/12/21/native-call-stack-profiling-3-3-2022-work-in-v8/).
+
+To compute the source code lines of a function, V8 knows the linear position of the function in the file. But, to find the source line, it needs to traverse the whole file to identify where each newline occurs. This is expensive.
+
+When requesting line information, V8 already knows how to cache the source line positions per-script. But the heap snapshot implementation was not using a cache!
```

Comment:
IIUC this is talking about `Script::InitLineEnds`. V8 cache this information when it knows that the script is being compiled in an active profiling session, and most importantly when it's allowed to store the line ends in the script (when heap allocation is allowed). Since heap snapshot can be generated anytime without alerting V8 in any way in advance, V8 could likely need this information in the middle of the heap explorer run, when it cannot save this information in the scripts since heap allocation is not allowed at that point. As a result V8 needs to recompute this information over and over again but cannot save it due to the unfortunate timing. The fix is to compute and cache all the line ends in advance when the heap allocation is still allowed.

It would be nicer to explain why this wasn't done before, and why the timing matters.

Reply:
In this case, I wanted to avoid going too much deep on the implementation details, and write more about the full process of iterating with profiling tools to solve the problem.

I agree on the interest of reviewing all this, to further explore improvements in the heap snapshot capture process. I.e. with the new changes, we actually allocate heap memory for line ends.

I think we should eventually move back to not allocating new line end information in heap snapshot capture (likely allocating a temporary line ends memory structure out of V8 heap that can be reused while heap snapshot is running, and is used only when line end information is not already available in the heap).

Reply:
I have been rereading this... and definitely there is a big difference between the detail for the first fix and the second fix. It even looks as I was in a hurry explaining the second one!

I will try to provide a more accurate description of what is fixed, mostly inspired in your first comment in this thread.